### DISCUSSION TITLE
What is the idomatic way to get and set structs from the UI to a Rust backend?

### DISCUSSION THREAD

**Discussion by alexeyabel** [slint-ui/slint#6165] [Q&A]
I struggle to find the idiomatic way to transfer several groups (structs) of values from and to the UI when using Rust. I know that I can read single values using
```rust
let ui: Main = ui_weak.upgrade().unwrap();
let myvalue = ui.get_myvalue();
```
but what is the idiomatic way to handle many values that can be grouped into structs (product type)?

Let's say I have a UI for a movie encoder and I have a set of values for the audio options and a set of values for the video options. Both sets are configured by the user, using widgets. Then the user clicks "Start Encoding". Now I need to get all these values into Rust. Let's just consider the simplified video options here:

I defined the struct in my .slint file:

```slint
export struct VideoSettingsUi {
    bitrate: int,
    quality: int,
    crop: bool,
}

export component Main inherits Window {
    callback encode_clicked(VideoSettingsUi);
}

// further below

Button {
     ...
    clicked => {
         root.encode_clicked({
              bitrate: bitrateSlider.value,
              quality: qualitySlider.value,
              crop: cropVideo.checked,
         });
    }
}
```

then in Rust I have something like

```rust
let ui_weak = ui_weak.clone();
        ui.on_encode_clicked(move |video_settings_ui: VideoSettingsUi| {
            let ui: Main = ui_weak.upgrade().unwrap();
```

1. Should I now also define an analog Rust struct? 
```rust
#[derive(Debug, Clone, PartialEq)]
pub struct VideoSettings {
    bitrate: u32,
    quality: u32,
    crop: bool,
}
```
and then implement a `From` trait to build a `VideoSettings` from `VideoSettingsUi`? 

2. Or can I just reuse whatever Slint generated for `VideoSettingsUi`? I also found `Model::slint` but it seems to be for something else entirely.
3. Or can I define something in Rust that is then injected as a struct in Slint without having to define `export struct VideoSettingsUi`?
4. Also, AFAIU `slint_build::compile("ui/main.slint").unwrap();` in build.rs will generate the code for `VideoSettingsUi` and `slint::include_modules!();` will include it. Is this the idiomatic way to do it? Are there others?
5. Does it make sense to define a root property such as `in-out property <VideoSettingsUi> videoSettings; `, to write the options into the UI through it on application startup?

I know that you have an open issue where a feature related to this is tracked. But if I develop the UI today, what do I do?

AFAICS, the examples in this repository never read/write a struct (multiple grouped values) between Rust and the UI. For beginners like me, it would be very helpful to have an authoritative answer that shows the idiomatic way.

**Comment by hunger** [comment]
You can take a look at how our Live-Preview works. The slint side of the API is in `tools/also/ui/API.slint`, the rust side in `tools/lsp/preview/ui.rs`.

You can definitely pass structs back and models back and forth. You can use the slint structs on the rust side, but you might prefer to have a different set of structs for rust as slint uses e.g. `slint::SharedString` over normal rust strings. So you might want to convert to something more rusty.

I hope this helps.

**Reply by alexeyabel** [reply]
Hi Tobias, thank you for the answer but I don't understand the content of the files you suggested. From what I can see, for example, the slint property `known-components` is set in ui.rs in `ui_set_known_components` and there is some complex stuff going on. In the end `result` is passed to `set_known_components()`. So there seems to be a correspondence between slints `ComponentListItem` and rusts `ComponentItem`, but I still don't understand what is really necessary for the conversion forth and back.

Is there maybe a simpler example somewhere in the code?

**Reply by hunger** [reply]
I have a data type in Rust and a somewhat different one on the Slint side. That is because I need to display components in the UI while I need to look up information on the Rust side. So there is a conversion happening, extracting and pre-processing information for display purposes.

It then sets the Slint-side struct using `set_known_components()`.
